### ISSUE TITLE
[EPIC] scale warehouse destination connectors to handle arbitrary number of streams

### ISSUE THREAD

**Issue by sherifnada** [airbytehq/airbyte#10260]
## Tell us about the problem you're trying to solve
Today destination warehouses or blob storages like S3 hold a constant-size, in-memory buffer for each stream it expects to see from the input catalog. For example, [this comment in the S3 code](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connectors/destination-s3/src/main/java/io/airbyte/integrations/destination/s3/util/S3StreamTransferManagerHelper.java#L18) illustrates that the more streams we have coming in, the more memory the connector must hold for buffers. 

This has two obvious problems. First is that it does not scale. Second, it's unnecessary because we often write files whose max size is no more than 1GB (because it's optimal for consuming warehouses). So, it doesn't really make sense overall that this works the way it does, and we stand to gain a lot of benefit from changing it. 

This current behavior also really hurts resource utilization when running a scaling airbyte instance, because you can't ever reliably predict how much memory a worker will need. 

## Describe the solution you’d like
I want the destination connectors to be able to work with a preset amount of memory e.g: I want to decide that if I have only 1.5GB to give the connector, then it should function correctly with that, maybe with a tradeoff of speed. But at least it shouldn't fail completely. Only hold space for a buffer if it is being used. We can even lazy-allocate space as records come in and release that space when the buffer is flushed, then allocate it again if new records come in for that buffer.

The most important destinations to do this for, in order, are: 

* Snowflake Internal staging
* S3 (because it is relied upon by GCS connector and therefore BigQuery connector. Also Redshift and Snowflake connectors, if the latter is using external staging to S3)

## Remaining pieces
Following the work on the first destination in Snowflake and S3, an abstraction was made to serialize records and compress them before buffering while monitoring for reasonable memory/file handlers limits.

This mechanism can now be replicated and deployed on all other java destinations by inheriting the same base classes while providing the proper `BlobStorageOperations` or `SqlOperations` to be interacted with.

The benefits can be shared for any destination that needs to buffer a batch of records before sending them instead of writing one by one.

# Warehouse & blob storage destinations 
It is a standard Warehouse best practice to use some staging areas where data is first buffered before being loaded to the final destinations. By nature, blob storage also follows the same requirement since they manipulate files.

* [ ] Snowflake destination
  * [x] Internal Staging & S3 external staging: #10866 
  * [x] GCS external Staging: #11428
  * [ ] Azure external Staging: #11429
* [x] S3 destination: #10591 / #11294
* [x] BigQuery destination: #11203
  * [x] BigQuery-denormalized
* [x] Redshift destination #11426
* [ ] Azure blob storage destination: #11421
* [ ] Databricks destination  #11424
* [x] GCS destination #11425

# Other destinations requiring buffering
The destinations in this section seem to already be using or needing some buffering mechanisms. These may be implemented on a case-by-case basis. Thus, refactoring them to follow the same pattern will reduce the maintenance cost of the connectors and unify the buffering options while getting memory consumptions under control in a common and more predictable manner:

* [ ] DynamoDB destination
* [ ] ElasticSearch destination
* [ ] MeiliSearch destination
* [ ] Pulsar destination
* [ ] RockSet destination

# Other destination that may benefit from buffering
The destination connectors below are sending records to the destination as they flow in. 
So, they may not have memory consumptions issues with large streams, but it might be useful to verify that batching in a buffer before sending a larger collection of records together may be beneficial:

* [ ] JDBC destination (these usually have some batch loading options and it's probably a good idea to use them too)
  * [ ] Clickhouse destination
  * [ ] MariaDbColumnStore destination
  * [ ] MsSql destination
  * [ ] MySql destination
  * [ ] Oracle destination
  * [ ] Postgres destination: use of `COPY` bulk loading is [recommended](https://www.postgresql.org/docs/9.2/populate.html)
* [ ] MongoDb destination
* [ ] MQTT destination

(* this may not be a completely exhaustive list, since we (airbyte or community) regularly add more destination connectors, so some destinations may not be listed in here, they may need to be added to the correct section or disregarded because buffering with that particular connector does not represent a benefit/issue)
(* Python connectors are not listed in this issue)

# Related areas
These are related areas that don't do so well with large batch or an arbitrary number of streams: 

* [ ] Normalization is being used by most of the data warehouse destinations in Airbyte: https://github.com/airbytehq/airbyte/issues/10977
* [ ] Sources: https://github.com/airbytehq/airbyte/issues/3439
* [ ] Source / discover: https://github.com/airbytehq/airbyte/issues/3943
* [ ] UI: https://github.com/airbytehq/airbyte/issues/3942
* [ ] UX: https://github.com/airbytehq/airbyte/issues/11389

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1202102171868563/1202126581516771) by [Unito](https://www.unito.io)

**Comment by tuliren** [comment]
- Snowflake Internal staging: https://github.com/airbytehq/airbyte/pull/10297
- S3: looks like it is addressed in https://github.com/airbytehq/airbyte/pull/9920

**Comment by tuliren** [comment]
The snowflake internal staging has stable memory usage for multiple streams.

One stream with tons of large records:

<img width="995" alt="Screen Shot 2022-02-16 at 15 03 33" src="https://user-images.githubusercontent.com/1933157/154372524-ead1323e-1a75-4c00-9855-40d1507ae53b.png">

10 streams with smaller load:
<img width="996" alt="Screen Shot 2022-02-16 at 15 04 11" src="https://user-images.githubusercontent.com/1933157/154372509-3a1c4f0a-e113-4a32-9ebb-4ea87c77db73.png">

I still need to figure out why it needs so much memory initially.

**Comment by tuliren** [comment]
One major contributor to the high memory usage is `ByteUtils#getSizeInBytesForUTF8CharSet`. This method creates a byte array from a string just to get a precise byte size. Along the way, it creates lots of byte arrays.

In practice, we only need an estimation of the string size. Since UTF-8 character is 1-4 bytes length, we can just use the string length times 4 to approximate the byte size. This will be an upper bound, which is even safer.

### Memory usage with `getSizeInBytesForUTF8CharSet`

<img width="658" alt="Screen Shot 2022-02-16 at 21 29 41" src="https://user-images.githubusercontent.com/1933157/154413185-2a8fad01-0013-4eb7-94eb-201851a35e7e.png">

<img width="1309" alt="Screen Shot 2022-02-16 at 21 30 19" src="https://user-images.githubusercontent.com/1933157/154413260-47ce4f9e-a730-44d2-8855-500ada26ed60.png">

We can see that:
- The max heap size is over 1.2 GB.
- Byte arrays take up 370 MB memory.

### Memory usage with simple estimation (string length x 4)

<img width="658" alt="Screen Shot 2022-02-16 at 21 41 18" src="https://user-images.githubusercontent.com/1933157/154413202-9a11f068-3ce9-4068-b98c-94ae96427ae6.png">

<img width="1310" alt="Screen Shot 2022-02-16 at 21 40 58" src="https://user-images.githubusercontent.com/1933157/154413288-4047efef-ecf4-4c4d-8d3a-a556f2060775.png">

The differences are:
- Max heap size is a little bit over 500 MB, as compared to 1.2+ GB.
- Byte arrays take up only 125 MB.

### Conclusion
Removing the `getSizeInBytesForUTF8CharSet` method calls immediately halves the memory usage.

**Comment by tuliren** [comment]
- The next idea is to further rely on estimation instead of precise calculation of the record message sizes.
- The proposal is that rather than serializing every record message, we can sample 1/10 of them and use the average size.
- The assumption behind this proposal is that each record message has roughly the same size.

**Comment by tuliren** [comment]
Sample the string size for every N records turn out to not have that big an effect.

The heap size is kept at 500 MB.

### Byte size (OOME with 500 MB heap size)
<img width="1603" alt="byte size every record oome" src="https://user-images.githubusercontent.com/1933157/154421621-acc149c4-6775-4e3f-a892-ab3f8bd321ab.png">

Calculating byte size for every record message results in OOME within one minute. This result proves that the frequent byte conversion is the root cause of the large memory footprint.

### String size (sample every record)
<img width="1605" alt="string size every record" src="https://user-images.githubusercontent.com/1933157/154421644-9c8cb4d4-b323-4be0-838a-0faa3a2f0752.png">

### String size (sample every 20 records)
<img width="1604" alt="string size every 20 records" src="https://user-images.githubusercontent.com/1933157/154421656-ceb8b186-03b3-4b2b-bc44-93337014701f.png">

### String size (sample every 100 records)
<img width="1609" alt="string size every 100 records" src="https://user-images.githubusercontent.com/1933157/154421663-53649d66-b7fc-4828-bfce-d6e163b7f0ab.png">

Performing one sample every 100 records slightly reduces memory usage. There are fewer number of memory usage peaks in the above chart.

**Comment by tuliren** [comment]
When setting the max heap size to 300 MB, the sampling rate has visible impact on heap usage. As the sampling rate decreases, there are fewer heap peaks, especially when the rate is reduced from sampling every record to every 20 records. The impact is not that significant when the rate changes from every 20 to every 100 records.

### String size (every record)
<img width="1603" alt="300mb string size every record" src="https://user-images.githubusercontent.com/1933157/154425470-25e615de-9bcd-4502-b76d-482854364219.png">

### String size (every 20 records)
<img width="1603" alt="300mb string size every 20 records" src="https://user-images.githubusercontent.com/1933157/154425486-94e020d2-2171-4579-91f4-f73d9a0a32f0.png">

### String size (every 100 records)
<img width="1601" alt="300mb string size every 100 records" src="https://user-images.githubusercontent.com/1933157/154425505-d6652d3e-b603-4593-a07a-2edc90fa8ff2.png">

### Conclusion
Will sample the string size every 20 records.

**Comment by tuliren** [comment]
300 MB seems like the minimum heap size. When setting a lower `xmx`, the connector went OOME within one minute:

- 200 MB max heap size:
  <img width="1603" alt="Screen Shot 2022-02-16 at 23 27 31" src="https://user-images.githubusercontent.com/1933157/154426758-6efe48bf-d8d0-4b30-9ba7-30f8aeb889d6.png">
- 256 MB max heap size:
  <img width="1603" alt="Screen Shot 2022-02-16 at 23 30 18" src="https://user-images.githubusercontent.com/1933157/154426771-81324bf7-20fe-4e9d-bb30-74941b65b20d.png">

**Comment by tuliren** [comment]
The S3 staging mode is not scalable. When syncing 50 streams, each with 1K records, the connector will encounter OOME with 500 MB max heap size.

Something is clearly wrong that the number of live threads just keeps going up. The majority of the heap is occupied by byte arrays. This mode requires more investigation.

<img width="1075" alt="Screen Shot 2022-02-23 at 07 20 34" src="https://user-images.githubusercontent.com/1933157/155349409-687c02e9-bd39-4b75-b296-aba16f5db965.png">

<img width="1074" alt="Screen Shot 2022-02-23 at 07 20 41" src="https://user-images.githubusercontent.com/1933157/155349439-9349f844-fb88-4fc2-af45-7da64964bc17.png">

<img width="1078" alt="Screen Shot 2022-02-23 at 07 20 50" src="https://user-images.githubusercontent.com/1933157/155349444-7603a4c6-632a-4613-a882-1418847888bd.png">

It does work with fewer streams (e.g. 3 streams, each with 10K records).

**Comment by tuliren** [comment]
## Summary
- ETA for S3 staging mode investigation and optimization: by the end of this week (Feb 25th).
- Remaining work
  - Add Sentry tracing for copy destination to facilitate investigation
  - Analyze the concurrent logic

**Comment by ChristopheDuong** [comment]
Following up on @tuliren's investigation:

> The S3 staging mode is not scalable. When syncing 50 streams, each with 1K records, the connector will encounter OOME with 500 MB max heap size.
> Something is clearly wrong that the number of live threads just keeps going up. The majority of the heap is occupied by byte arrays. This mode requires more investigation.

PR #10866 solves the non-scalable aspects that were discovered so far with the S3 staging mode. Below numbers are run with the new S3 staging mode, syncing 200 streams, each with 2K records (representing 4.24GB of data in total) running with 200MB max heap size.

![Screenshot 2022-03-21 at 10 25 13](https://user-images.githubusercontent.com/7718175/159234360-74d9cf6a-5487-45b4-8fe0-410d3b34909c.png)
![Screenshot 2022-03-21 at 10 25 22](https://user-images.githubusercontent.com/7718175/159234378-98f83526-4eb5-4fbf-be2d-11c4d8ed3f99.png)

[S3-staging-200_streams_1000_records_1024mb_gzip_disk_buffer.log](https://github.com/airbytehq/airbyte/files/8314330/S3-staging-200_streams_1000_records_1024mb_gzip_disk_buffer.log)

**Comment by ChristopheDuong** [comment]
> PR https://github.com/airbytehq/airbyte/pull/10866 solves the non-scalable aspects that were discovered so far with the S3 staging mode. Below numbers are run with the new S3 staging mode, syncing 200 streams, each with 2K records (representing 4.24GB of data in total) running with 200MB max heap size.

However, these are not the optimal / recommended max heap sizes to run the new staging mode with.

When the same connector is run with the same configuration, but with a larger amount of data per stream, if we keep the max heap size of 200MB, then we encounter OOME...

For example, let's study a scenario where N stream with 100K records is representing 750-800MB of uncompressed data per stream. When compressed, it is a little more than 200MB compressed (gzip) of data per stream.

Even though the data is now being buffered on-disk per stream, when a flush event occurs for a particular stream, the compressed data for that stream is fully loaded into memory (with extra memory overhead as described [here](https://github.com/alexmojaki/s3-stream-upload/blob/e13637e8fc9936f1aa3deb94ccf3678ca5bc8ddc/src/main/java/alex/mojaki/s3upload/StreamTransferManager.java#L97)). (note that the connector can buffer multiple streams at once in different file buffers on disk, but for the moment, it will load only one file at a time when triggered for uploading to staging)

![Screenshot 2022-03-21 at 10 10 40](https://user-images.githubusercontent.com/7718175/159236698-f371c9cc-4d06-4edd-a8cc-f5aec4cb1883.png)

As a conclusion, it is better to run with at least somewhere around 600MB max heap size (with the current implementation of the buffering configured with 200MB maximum of compressed data in the buffer)

**Comment by ChristopheDuong** [comment]
> it is better to run with at least somewhere around 600MB max heap size + 1GB disk storage

Destination snowflake with Internal Staging does not have the same memory requirement because it is not relying on the `StreamTransferManager` class. So it does not actually load the buffered file into memory in order to upload it to staging.

## Scenario 1 of 4.24GB over 200 streams and 2K records (5MB compressed) per stream

![Screenshot 2022-03-21 at 16 09 52](https://user-images.githubusercontent.com/7718175/159291048-3668dbbc-e3ea-4fd5-a51a-a7176bef39a6.png)

## Scenario 2 of 15GB over 15 streams and 100K records (283MB compressed) per stream

### Memory consumption when buffering/processing the first 200K records on 2 stream:
![Screenshot 2022-03-21 at 15 59 17](https://user-images.githubusercontent.com/7718175/159290154-67f7e47d-1168-40bf-be8f-498e44e2bb3c.png)

### Memory consumption in a "stable" state when reaching 10+GB of data transferred (1000+K records on 10+ streams)

**Comment by ChristopheDuong** [comment]
> a relatively large dataset (~6m rows, so nothing too wild).
> I also see the memory usage by that pod hit the node's capacity just before it dies. Seems like it's related to this general issue, 

How are the 6M rows split on how many tables?
If it's coming from a large number of tables. then yes it might be related to what is being targeted by this epic issue and the one you linked.

A workaround would be to split your connection and select only a few tables because the connector is allocating a large buffer for each stream all at once.

For ETA, I don't know. maybe @grishick  can answer that part.

**Comment by olivermeyer** [comment]
> How are the 6M rows split on how many tables?

It's actually just one table, so perhaps it's another bug. I'll do a bit more digging and open a new issue.

**Comment by ChristopheDuong** [comment]
> > How are the 6M rows split on how many tables?
> 
> It's actually just one table, so perhaps it's another bug. Perhaps I should open a new issue?

Yes, please!

- Can you specify how much memory do you allocate to the worker node running the connector?
- Can you try the destination-s3 connector to see if the sync fails there too? (if that works, then while waiting on a fix on destination-redshift, you could then manually load the data from s3 to redshift)

**Comment by sherifnada** [comment]
It could be a wide row issue potentially
